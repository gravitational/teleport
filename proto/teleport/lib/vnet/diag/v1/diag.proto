// Teleport
// Copyright (C) 2025 Gravitational, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

syntax = "proto3";

package teleport.lib.vnet.diag.v1;

import "google/protobuf/timestamp.proto";

option go_package = "github.com/gravitational/teleport/gen/proto/go/teleport/lib/vnet/diag/v1;diagv1";

// Report represents the attempts at running individual checks. It also includes general information
// about the network stack managed by VNet. It assumes that each individual check as well as getting
// info about the network stack can fail.
message Report {
  // created_at is the UTC timestamp at which the report was generated.
  google.protobuf.Timestamp created_at = 1;
  NetworkStackAttempt network_stack_attempt = 2;
  repeated CheckAttempt checks = 3;
}

// NetworkStackAttempt represents the attempt at getting information about the network stack managed
// by VNet.
message NetworkStackAttempt {
  CheckAttemptStatus status = 1;
  // error is present if status is CHECK_ATTEMPT_STATUS_ERROR.
  string error = 2;
  NetworkStack network_stack = 3;
}

// NetworkStack describes the network stack managed by VNet.
message NetworkStack {
  // interface_name is the name of the interface set up and used by VNet.
  string interface_name = 1;
  // ipv4_cidr_ranges are all the possible ranges under which VNet is going to assign IPv4 addresses
  // for apps. The first IP of the first range is used for the TUN device.
  // Each root cluster can specify its own CIDR range to be used for apps within that profile.
  repeated string ipv4_cidr_ranges = 2;
  // ipv6_prefix is the IPv6 prefix under which VNet creates IPv6 addresses for apps and its DNS
  // server.
  string ipv6_prefix = 3;
  // dns_zones lists domains for which DNS queries are supposed to be captured by VNet.
  repeated string dns_zones = 4;
}

// CheckAttempt describes the attempt at running a particular diagnostic check. If it succeeds
// (status is CHECK_ATTEMPT_STATUS_OK), check_report can be inspected to see if the check has found
// any issues.
//
// For example, a check that inspects network routes can succeed (CHECK_ATTEMPT_STATUS_OK) and it
// might or might not find conflicting routes (CHECK_REPORT_STATUS_ISSUES_FOUND or
// CHECK_REPORT_STATUS_OK). But it can also fail to run (CHECK_ATTEMPT_STATUS_ERROR) because the
// syscall to list routes has failed.
message CheckAttempt {
  // status represents the status of an attempt at running a particular diagnostic check. This is
  // not the same as the status of CheckReport.
  //
  CheckAttemptStatus status = 1;
  // error is present if the check failed to run (status is CHECK_ATTEMPT_STATUS_ERROR).
  string error = 2;
  // check_report is the output of a particular check.
  //
  // If check failed to run (status is CHECK_ATTEMPT_STATUS_ERROR), the report oneof in check_report
  // is set to a specific member while the member itself is empty. This means that a particular
  // CheckAttempt can be distinguished from other attempts describing other checks even if the check
  // failed to run.
  CheckReport check_report = 3;
  // commands are the outputs from additional diagnostic commands executed by the diagnostic
  // check. They are meant to help inspect the general state of the OS related to the given check.
  // Unless a callsite specifically requests commands to be skipped, commands are present even if
  // status is CHECK_ATTEMPT_STATUS_ERROR, as they are useful even if the check failed.
  repeated CommandAttempt commands = 4;
}

// CheckAttemptStatus describes whether CheckAttempt finished successfully. This is different from
// CheckReportStatus, which describes whether a successful attempt at running a check has found any
// issues. See the comment for CheckAttempt for an example.
enum CheckAttemptStatus {
  CHECK_ATTEMPT_STATUS_UNSPECIFIED = 0;
  // CHECK_ATTEMPT_STATUS_OK indicates that the check ran successfully.
  CHECK_ATTEMPT_STATUS_OK = 1;
  // CHECK_ATTEMPT_STATUS_ERROR indicates that the check failed to run.
  CHECK_ATTEMPT_STATUS_ERROR = 2;
}

// CheckReport is the output of a successful attempt at running a particular check.
message CheckReport {
  // status indicates if the check has found any issues. This is so that a callsite operating on
  // a CheckReport can understand the outcome of the check without having to understand the
  // semantics of the output included under report.
  CheckReportStatus status = 1;

  oneof report {
    // route_conflict reports whether there are routes that might conflict with routes set up by
    // VNet.
    RouteConflictReport route_conflict_report = 2;
  }
}

// CheckReportStatus describes the outcome of a successful attempt at running a check.
enum CheckReportStatus {
  CHECK_REPORT_STATUS_UNSPECIFIED = 0;
  // CHECK_REPORT_STATUS_OK indicates that the check has not found any issues.
  CHECK_REPORT_STATUS_OK = 1;
  // CHECK_REPORT_STATUS_ISSUES_FOUND indicates that the check has found at least one issue.
  CHECK_REPORT_STATUS_ISSUES_FOUND = 2;
}

// CommandAttempt describes the attempt at running a particular command associated with a diagnostic
// check.
message CommandAttempt {
  CommandAttemptStatus status = 1;
  // error is present if status is COMMAND_ATTEMPT_STATUS_ERROR.
  string error = 2;
  // command shows which command was executed along with its arguments, e.g., "netstat -rn -f inet".
  string command = 3;
  // output is stdout from the command if status is COMMAND_ATTEMPT_STATUS_OK.
  string output = 4;
}

// CommandAttemptStatus describes the status of CommandAttempt.
enum CommandAttemptStatus {
  COMMAND_ATTEMPT_STATUS_UNSPECIFIED = 0;
  COMMAND_ATTEMPT_STATUS_OK = 1;
  COMMAND_ATTEMPT_STATUS_ERROR = 2;
}

// RouteConflictReport describes conflicting routes found by RouteConflictDiag.
message RouteConflictReport {
  repeated RouteConflict route_conflicts = 1;
}

// RouteConflict describes a conflict between a route set up by a 3rd-party app where the
// destination overlaps with a destination in a route set up by VNet.
message RouteConflict {
  // dest is the destination of the conflicting route.
  string dest = 1;
  // vnet_dest is the destination of a VNet route that Dest overlaps with.
  string vnet_dest = 2;
  // interface_name is the name of the interface the route uses, e.g. "utun4".
  string interface_name = 3;
  // interface_app may contain the name of the application responsible for setting up the interface.
  // At the moment, the only source of this information is NetworkExtension description included in
  // the output of `ifconfig -v <interface name>`. Not all VPN applications use this framework, so
  // it's likely to be empty.
  string interface_app = 4;
}
