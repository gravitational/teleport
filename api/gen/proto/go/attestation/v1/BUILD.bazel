load("@io_bazel_rules_go//go:def.bzl", "go_library")

go_library(
    name = "attestation",
    srcs = ["attestation.pb.go"],
    importpath = "github.com/gravitational/teleport/api/gen/proto/go/attestation/v1",
    visibility = ["//visibility:public"],
    deps = ["@com_github_gogo_protobuf//proto"],
)
