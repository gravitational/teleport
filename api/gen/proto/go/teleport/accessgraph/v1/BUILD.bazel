load("@io_bazel_rules_go//go:def.bzl", "go_library")

go_library(
    name = "accessgraph",
    srcs = [
        "authorized_key.pb.go",
        "private_key.pb.go",
        "secrets_service.pb.go",
        "secrets_service_grpc.pb.go",
    ],
    importpath = "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1",
    visibility = ["//visibility:public"],
    deps = [
        "//api/gen/proto/go/teleport/devicetrust/v1:devicetrust",
        "//api/gen/proto/go/teleport/header/v1:header",
        "@org_golang_google_grpc//:grpc",
        "@org_golang_google_grpc//codes",
        "@org_golang_google_grpc//status",
        "@org_golang_google_protobuf//reflect/protoreflect",
        "@org_golang_google_protobuf//runtime/protoimpl",
    ],
)
