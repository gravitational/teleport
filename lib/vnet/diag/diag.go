// Teleport
// Copyright (C) 2025 Gravitational, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package diag

import (
	"context"
	"errors"
	"fmt"
	"os/exec"
	"strings"
	"unicode/utf8"

	"github.com/gravitational/trace"
	"github.com/jonboulle/clockwork"
	"google.golang.org/protobuf/types/known/timestamppb"

	"github.com/gravitational/teleport"
	diagv1 "github.com/gravitational/teleport/gen/proto/go/teleport/lib/vnet/diag/v1"
	logutils "github.com/gravitational/teleport/lib/utils/log"
)

var log = logutils.NewPackageLogger(teleport.ComponentKey, teleport.Component("vnet", "diag"))

// DiagCheck is an individual diag check run by [GenerateReport].
type DiagCheck interface {
	// Run performs the check.
	Run(context.Context) (*diagv1.CheckReport, error)
	// Commands returns commands accompanying the check which are supposed to help inspect the state of
	// the OS relevant to the given check even if the check itself fails.
	Commands(context.Context) []*exec.Cmd
	// EmptyCheckReport returns an empty version of [diagv1.CheckReport] belonging to this DiagCheck.
	// If Run fails, it's used to set the correct kind of [diagv1.CheckReport] on [diagv1.CheckAttempt].
	EmptyCheckReport() *diagv1.CheckReport
}

// ReportPrerequisites are items needed by [GenerateReport].
type ReportPrerequisites struct {
	Clock               clockwork.Clock
	NetworkStackAttempt *diagv1.NetworkStackAttempt
	DiagChecks          []DiagCheck
	// SkipCommands controls whether the report returned from [GenerateReport] is going to include
	// extra commands accompanying each diagnostic check. Useful in contexts where there's no place to
	// display output of those commands.
	SkipCommands bool
}

func (rp *ReportPrerequisites) check() error {
	if rp.Clock == nil {
		return trace.BadParameter("missing clock")
	}

	if rp.NetworkStackAttempt == nil {
		return trace.BadParameter("missing network stack attempt")
	}

	if len(rp.DiagChecks) == 0 {
		return trace.BadParameter("no diag checks provided")
	}

	return nil
}

// GenerateReport generates a report using the output of the checks provided through [rp].
func GenerateReport(ctx context.Context, rp ReportPrerequisites) (*diagv1.Report, error) {
	if err := rp.check(); err != nil {
		return nil, trace.Wrap(err)
	}

	report := &diagv1.Report{}
	report.CreatedAt = timestamppb.New(rp.Clock.Now().UTC())
	report.NetworkStackAttempt = rp.NetworkStackAttempt

	for _, diagCheck := range rp.DiagChecks {
		checkAttempt := runCheck(ctx, diagCheck, rp.SkipCommands)

		report.Checks = append(report.Checks, checkAttempt)
	}

	return report, nil
}

func runCheck(ctx context.Context, diagCheck DiagCheck, skipCommands bool) *diagv1.CheckAttempt {
	attempt := &diagv1.CheckAttempt{}

	report, err := diagCheck.Run(ctx)
	if err != nil {
		attempt.Status = diagv1.CheckAttemptStatus_CHECK_ATTEMPT_STATUS_ERROR
		attempt.Error = err.Error()
		// In case of an error, CheckReport needs to be set to an empty value. Otherwise it'd be
		// impossible to identify the type of a failed check.
		attempt.CheckReport = diagCheck.EmptyCheckReport()
	} else {
		attempt.Status = diagv1.CheckAttemptStatus_CHECK_ATTEMPT_STATUS_OK
		attempt.CheckReport = report
	}

	if !skipCommands {
		for _, cmd := range diagCheck.Commands(ctx) {
			attempt.Commands = append(attempt.Commands, runCommand(cmd))
		}
	}

	return attempt
}

func runCommand(cmd *exec.Cmd) *diagv1.CommandAttempt {
	command := strings.Join(cmd.Args, " ")

	output, err := cmd.Output()
	if err != nil {
		var exitError *exec.ExitError
		errMessage := err.Error()
		if errors.As(err, &exitError) {
			stderr := string(exitError.Stderr)
			if stderr != "" && utf8.Valid(exitError.Stderr) {
				errMessage = fmt.Sprintf("%s\n%s", errMessage, stderr)
			}
		}

		return &diagv1.CommandAttempt{
			Status:  diagv1.CommandAttemptStatus_COMMAND_ATTEMPT_STATUS_ERROR,
			Error:   errMessage,
			Command: command,
		}
	}

	// A protobuf string must contain valid UTF-8 or 7-bit ASCII text .
	if !utf8.Valid(output) {
		return &diagv1.CommandAttempt{
			Status:  diagv1.CommandAttemptStatus_COMMAND_ATTEMPT_STATUS_ERROR,
			Error:   "command output contains text that is not UTF-8 encoded",
			Command: command,
		}
	}

	return &diagv1.CommandAttempt{
		Status:  diagv1.CommandAttemptStatus_COMMAND_ATTEMPT_STATUS_OK,
		Command: command,
		Output:  string(output),
	}
}
