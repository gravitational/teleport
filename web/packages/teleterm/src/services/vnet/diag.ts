/**
 * Teleport
 * Copyright (C) 2025 Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { displayDateTime } from 'design/datetime';
import { Timestamp } from 'gen-proto-ts/google/protobuf/timestamp_pb';
import * as diag from 'gen-proto-ts/teleport/lib/vnet/diag/v1/diag_pb';

import { reportOneOfIsRouteConflictReport } from 'teleterm/helpers';

export const hasReportFoundIssues = (report: diag.Report): boolean =>
  report.checks.some(
    checkAttempt =>
      checkAttempt.status === diag.CheckAttemptStatus.OK &&
      checkAttempt.checkReport.status === diag.CheckReportStatus.ISSUES_FOUND
  );

export const getReportFilename = (report: diag.Report) => {
  const createdAt = displayDateTime(Timestamp.toDate(report.createdAt));
  // Colons are best avoided on macOS and forbidden on Windows.
  // https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations
  // Spaces are removed as well to avoid issues with encoding if the user uploads the file and makes
  // it accessible over a URL that includes the filename.
  const sanitizedCreatedAt = createdAt
    .replaceAll(' ', '_')
    .replaceAll(':', '-');

  return `vnet_diag_report_${sanitizedCreatedAt}.txt`;
};

/**
 * reportToText serializes the report into text that can be shared by the user. It was written
 * primarily with Zendesk and Slack in mind. As of February 2025, Slack doesn't support Markdown
 * tables, but Zendesk does.
 *
 * Still, the text should be light on Markdown as the user might post it to a platform that
 * doesn't support Markdown. For example, we should refrain from using <details>.
 */
export function reportToText(report: diag.Report): string {
  const createdAt = Timestamp.toDate(report.createdAt);
  const localCreatedAt = displayDateTime(createdAt);
  const utcCreatedAt = createdAt.toUTCString();
  const networkStack = networkStackAttemptToText(report.networkStackAttempt);
  const checkReports = report.checks.map(checkAttemptToText).join('\n\n');

  return `VNet Diagnostic Report

Created at: ${localCreatedAt} (${utcCreatedAt})
${networkStack}

${checkReports}\n`;
}

function networkStackAttemptToText(
  networkStackAttempt: diag.NetworkStackAttempt
): string {
  const { networkStack } = networkStackAttempt;

  return networkStackAttempt.status === diag.CheckAttemptStatus.ERROR
    ? `Network details could not be determined:
${networkStackAttempt.error}`
    : `Network interface: ${networkStack.interfaceName}
IPv4 CIDR ranges: ${networkStack.ipv4CidrRanges.join(', ')}
IPv6 prefix: ${networkStack.ipv6Prefix}
DNS zones: ${networkStack.dnsZones.join(', ')}`;
}

function checkAttemptToText(checkAttempt: diag.CheckAttempt): string {
  const reportOneof = checkAttempt.checkReport.report.oneofKind;
  const displayDetails = reportOneofToDisplayDetails[reportOneof];

  if (!displayDetails) {
    return `Cannot display the result from an unsupported check ${reportOneof}.`;
  }

  let checkSummary: string;
  if (checkAttempt.status === diag.CheckAttemptStatus.ERROR) {
    checkSummary = `Failed to ${displayDetails.errorTitle}:
${checkAttempt.error}`;
  } else {
    checkSummary = displayDetails.reportToText(checkAttempt.checkReport);
  }

  const commandSummaries = checkAttempt.commands.map(commandAttempt =>
    commandAttempt.status === diag.CommandAttemptStatus.ERROR
      ? `Ran into an error when executing ${commandAttempt.command}:
${commandAttempt.error}`
      : `\`\`\`
$ ${commandAttempt.command}
${commandAttempt.output}
\`\`\``
  );

  return `---
${checkSummary}

${commandSummaries.join('\n\n')}`;
}

const reportOneofToDisplayDetails: Record<
  diag.CheckReport['report']['oneofKind'],
  {
    reportToText: (checkReport: diag.CheckReport) => string;
    errorTitle: string;
  }
> = {
  routeConflictReport: {
    errorTitle: 'inspect network routes',
    reportToText: routeConflictReportToText,
  },
};

function routeConflictReportToText({
  report,
  status,
}: diag.CheckReport): string {
  if (!reportOneOfIsRouteConflictReport(report)) {
    return '';
  }

  if (status === diag.CheckReportStatus.OK) {
    return '✅ There are no network routes in conflict with VNet.';
  }

  const tableRows = report.routeConflictReport.routeConflicts
    .map(
      routeConflict =>
        `| ${routeConflict.dest} | ${routeConflict.vnetDest} | ${routeConflict.interfaceName} | ${routeConflict.interfaceApp || 'unknown'} |`
    )
    .join('\n');

  return `⚠️ There are network routes in conflict with VNet.

| Conflicting destination | VNet destination | Interface | Set up by |
| ----------------------- | ---------------- | --------- | --------- |
${tableRows}`;
}
