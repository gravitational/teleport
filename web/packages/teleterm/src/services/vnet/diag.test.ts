/**
 * Teleport
 * Copyright (C) 2025 Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import * as diag from 'gen-proto-ts/teleport/lib/vnet/diag/v1/diag_pb';

import { reportToText } from './diag';
import {
  makeCheckAttempt,
  makeCheckReport,
  makeCommandAttempt,
  makeReport,
  makeRouteConflict,
} from './testHelpers';

describe('reportToText', () => {
  it('converts report correctly', () => {
    const checkReport = makeCheckReport({
      status: diag.CheckReportStatus.ISSUES_FOUND,
    });
    checkReport.report = {
      oneofKind: 'routeConflictReport',
      routeConflictReport: {
        routeConflicts: [
          makeRouteConflict({
            dest: '100.64.0.0/10',
            vnetDest: '100.64.0.1',
            interfaceName: 'utun5',
            interfaceApp: 'VPN: Foobar',
          }),
          makeRouteConflict({
            dest: '0.0.0.0/1',
            vnetDest: '100.64.0.0/10',
            interfaceName: 'utun6',
            interfaceApp: '',
          }),
        ],
      },
    };
    const report = makeReport({
      checks: [
        makeCheckAttempt({
          checkReport,
          commands: [makeCommandAttempt()],
        }),
      ],
    });

    const actualText = reportToText(report);
    expect(actualText).toMatchSnapshot();
    // Verify that the text ends with a newline.
    expect(actualText.endsWith('\n')).toBe(true);
  });
});
