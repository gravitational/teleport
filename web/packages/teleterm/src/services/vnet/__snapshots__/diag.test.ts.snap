// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reportToText converts report correctly 1`] = `
"VNet Diagnostic Report

Created at: 2024-11-23 13:27:48 (Sat, 23 Nov 2024 13:27:48 GMT)
Network interface: utun4
IPv4 CIDR ranges: 100.64.0.0/10
IPv6 prefix: fdff:fd74:46c0::
DNS zones: teleport.example.com, company.test

---
⚠️ There are network routes in conflict with VNet.

| VNet destination | Conflicting destination | Interface | Set up by |
| ---------------- | ----------------------- | --------- | --------- |
| 100.64.0.1 | 100.64.0.0/10 | utun5 | VPN: Foobar |
| 100.64.0.0/10 | 0.0.0.0/1 | utun6 | unknown |

\`\`\`
$ netstat -rn -f inet
Routing tables

Internet:
Destination        Gateway            Flags               Netif Expire
default            192.168.1.1        UGdScg                en0       
default            link#23            UCSIg           bridge100      !
default            link#25            UCSIg               utun4       
100.64/10          link#25            UCS                 utun4       
100.64.0.1         100.64.0.1         UH                  utun5       
100.87.112.117     100.87.112.117     UH                  utun4       
100.100.100.100/32 link#25            UCS                 utun4       
100.100.100.100    link#25            UHWIi               utun4       

\`\`\`
"
`;
